js/src/logger.ts

```diff
@@ -997,18 +997,32 @@ export class Logger<IsAsyncFlush extends boolean> implements Exportable {
   ): PromiseUnless<IsAsyncFlush, R> {
     const { setCurrent, ...argsRest } = args ?? {};
     const span = this.startSpan(argsRest);
-    const ret = runFinally(
-      () => {
-        if (setCurrent ?? true) {
-          return withCurrent(span, callback);
-        } else {
-          return callback(span);
-        }
-      },
-      () => span.end(),
-    );
+
+    let ret;
+    try {
```

Comment:
I think it would be cleaner to add a `runCatch` utility analogous to our `runFinally` util and use that. Something like
```
// In util.ts

// Given a function to execute and a catch handler, runs the function and
// executes the catch handler either synchronously if `f` is a synchronous
// function or chained if `f` returns a Promise.
export function runCatch<R>(f: () => R, catchF: (e: unknown) => R): R {
  try {
    const ret = f();
    if (ret instanceof Promise) {
      return (ret as any).catch(catchF) as R;
    } else {
      return ret;
    }
  } catch (e) {
    return catchF(e);
  }
}
```

Then here we could write:
```
    const ret = runFinally(
        () => runCatch(
          () => {
            if (setCurrent ?? true) {
              return withCurrent(span, callback);
            } else {
              return callback(span);
            }
          },
          (e) => {
            logError(span, e);
            throw e;
          }),
        () => span.end());
```

And leave the rest the same.

Reply:
updated (but made `runFinally` accept a catch clause, because everywhere we use it, we want to do this)